### PR TITLE
mock-array: modify constraints to model clock arriving earlier

### PR DESCRIPTION
**Pull request by oharboe**
The goal is to model the performance of mock-array when it is used in a context where the clock network latency of the top level that contains mock-array is greater than the mock-array.

CTS will once https://github.com/The-OpenROAD-Project/OpenROAD/issues/3759 is implemented account for the macro clock network latency, which will improve the `report_checks -path_delay max` as ca. 100ps is added to the max path due to Element clock network latency.

This is based on the suggestion from @rovinski : https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1281#issuecomment-1656568687

Timing is met with a 500ps clock period:

```
>>> report_checks -path_delay max
Startpoint: ces_7_0 (rising edge-triggered flip-flop clocked by clock)
Endpoint: _187_ (falling edge-triggered flip-flop clocked by clock')
Path Group: clock
Path Type: max

  Delay    Time   Description
---------------------------------------------------------
   0.00    0.00   clock clock (rise edge)
 281.93  281.93   clock network delay (propagated)
   0.00  281.93 ^ ces_7_0/clock (Element)
 163.75  445.68 v ces_7_0/io_lsbOuts_7 (Element)
  36.29  481.97 v ces_7_1/io_lsbOuts_6 (Element)
  36.47  518.44 v ces_7_2/io_lsbOuts_5 (Element)
  36.29  554.73 v ces_7_3/io_lsbOuts_4 (Element)
  36.47  591.20 v ces_7_4/io_lsbOuts_3 (Element)
  36.29  627.49 v ces_7_5/io_lsbOuts_2 (Element)
  36.29  663.78 v ces_7_6/io_lsbOuts_1 (Element)
  37.14  700.92 v ces_7_7/io_lsbOuts_0 (Element)
   0.23  701.14 v _187_/D (DFFLQNx2_ASAP7_75t_R)
         701.14   data arrival time

 500.00  500.00   clock clock' (fall edge)
 255.31  755.31   clock network delay (propagated)
   1.37  756.68   clock reconvergence pessimism
         756.68 v _187_/CLK (DFFLQNx2_ASAP7_75t_R)
  -6.72  749.96   library setup time
         749.96   data required time
---------------------------------------------------------
         749.96   data required time
        -701.14   data arrival time
---------------------------------------------------------
          48.82   slack (MET)


>>> report_checks -path_delay min
Startpoint: ces_4_7 (rising edge-triggered flip-flop clocked by clock)
Endpoint: ces_5_7 (rising edge-triggered flip-flop clocked by clock)
Path Group: clock
Path Type: min

  Delay    Time   Description
---------------------------------------------------------
   0.00    0.00   clock clock (rise edge)
 229.61  229.61   clock network delay (propagated)
   0.00  229.61 ^ ces_4_7/clock (Element)
 145.09  374.70 ^ ces_4_7/io_outs_up[25] (Element)
   0.03  374.74 ^ ces_5_7/io_ins_up[25] (Element)
         374.74   data arrival time

   0.00    0.00   clock clock (rise edge)
 273.17  273.17   clock network delay (propagated)
  -1.37  271.80   clock reconvergence pessimism
         271.80 ^ ces_5_7/clock (Element)
  89.58  361.38   library hold time
         361.38   data required time
---------------------------------------------------------
         361.38   data required time
        -374.74   data arrival time
---------------------------------------------------------
          13.36   slack (MET)
```

Some more detail from Element:

```
>>> report_checks -path_delay max -to io_lsbOuts_7 -format full_clock
Startpoint: _749_ (falling edge-triggered flip-flop clocked by clock')
Endpoint: io_lsbOuts_7 (output port clocked by clock)
Path Group: clock
Path Type: max

  Delay    Time   Description
---------------------------------------------------------
   0.00    0.00   clock clock' (fall edge)
   0.00    0.00   clock source latency
   6.81    6.81 ^ clock (in)
  21.55   28.37 ^ clkbuf_0_clock/Y (BUFx4_ASAP7_75t_R)
  24.27   52.63 ^ clkbuf_1_0__f_clock/Y (BUFx4_ASAP7_75t_R)
  51.78  104.42 ^ clkbuf_leaf_1_clock/Y (BUFx4_ASAP7_75t_R)
  18.33  122.74 v net499_21/Y (INVx3_ASAP7_75t_R)
   0.03  122.78 v _749_/CLK (DFFLQNx2_ASAP7_75t_R)
  48.41  171.19 v _749_/QN (DFFLQNx2_ASAP7_75t_R)
  13.66  184.84 ^ _496_/Y (INVx3_ASAP7_75t_R)
  21.84  206.68 ^ output271/Y (BUFx2_ASAP7_75t_R)
   0.25  206.94 ^ io_lsbOuts_7 (out)
         206.94   data arrival time

 500.00  500.00   clock clock (rise edge)
   0.00  500.00   clock network delay (propagated)
 -20.00  480.00   clock uncertainty
   0.00  480.00   clock reconvergence pessimism
  75.00  555.00   output external delay
         555.00   data required time
---------------------------------------------------------
         555.00   data required time
        -206.94   data arrival time
---------------------------------------------------------
         348.06   slack (MET)
```

Clock Tree Viewer from Element:

image

### CODE REVIEW

File: flow/designs/asap7/mock-array/Element/constraints.sdc [hunk 1]
```diff
@@ -4,34 +4,20 @@ set cols [expr {[info exists ::env(MOCK_ARRAY_COLS)] ? $::env(MOCK_ARRAY_COLS) :
...
 set clk_port [get_ports $clk_port_name]
 create_clock -period $clk_period -waveform [list 0 [expr $clk_period / 2]] -name $clk_name $clk_port
 set_clock_uncertainty -setup 20.0 [get_clocks $clk_name]
 set_clock_uncertainty -hold 20.0 [get_clocks $clk_name]
 
+set clock_network_latency 100
+
 # io_ins_x -> REG_x in neighbouring element or just outside of the array
-set_input_delay -clock $clk_name [expr $clk_period * 0.05] [get_ports {io_ins_*}]
+set_input_delay -clock $clk_name [expr $clock_network_latency + $clk_period * 0.05] [get_ports {io_ins_* io_lsbIns_4}]
```
Comment (maliberty) [inline flow/designs/asap7/mock-array/Element/constraints.sdc]: Why just io_lsbIns_4 ?

Comment (oharboe) [reply flow/designs/asap7/mock-array/Element/constraints.sdc]: will remove, belongs to a different PR

Comment (mynnguyen) [reply flow/designs/asap7/mock-array/Element/constraints.sdc]: Is the question about what if clock latency greater than clock period and how to constraints IO paths accordingly?
One way to compensate for "actual clock tree delay/latency" is the use of virtual clock to constraint the IO.
(It has similar effect as suggestion from @rovinski, but I guess his is using real clock to constraint the IO)

Comment (oharboe) [reply flow/designs/asap7/mock-array/Element/constraints.sdc]: > Is the question about what if clock latency greater than clock period and how to constraints IO paths accordingly?

Not really, I think that is a separate concern that I did not intend to address in this PR.

My thinking is that clock can be inserted *earlier* when mock-array is used as part of a macro where the clock network latency is significantly greater than the clock network latency of mock-array as long as the clock tree of the macro that contains mock-array is balanced approperiately.

This is what I'm trying to model here and this allows me to reduce the clock period of the mock-array as well as getting rid of all the hold cells.


> One way to compensate for "actual clock tree delay/latency" is the use of virtual clock to constraint the IO. (It has similar effect as suggestion from @rovinski, but I guess his is using real clock to constraint the IO)

Is there, at this point, a reason to prefer a different approach to the one used here?

I have never used virtual clocks so I don't know how to articulate that alternative solution. Perhaps that could be a future refinement on the approach?

Comment (oharboe) [reply flow/designs/asap7/mock-array/Element/constraints.sdc]: @mynnguyen I've asked for a second review to follow up on the above.

**Comment by rovinski** [comment]
@oharboe does the mock array design have any flip flops at the top level? Or are they all inside the Element? Even if https://github.com/The-OpenROAD-Project/OpenROAD/issues/3759 is implemented, it won't change results if all macros are the same and there are no flip flops. The change would be to help _mismatches_ for different macros. For example, the clock "tree" in flip flops are extremely small, maybe a picosecond or less. The clock tree of "Element" is much larger, I think I saw a couple hundred picoseconds, so it would help there. But if there are no flip flops and only element macros, then there is no mismatch between identical instances of macros.

**Comment by oharboe** [comment]
> @oharboe does the mock array design have any flip flops at the top level? 

Yes.

**Pull request closed without merging**